osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using osu.Framework.Bindables;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Configuration;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Overlays.Settings;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Rulesets.Judgements;
+
+namespace osu.Game.Rulesets.Mods
+{
+    public class ModAccuracyChallenge : ModFailCondition
+    {
+        public override string Name => "Accuracy Challenge";
+        public override string Acronym => "AC";
+        public override string Description => "Fail the map if you don't maintain a certain accuracy.";
+        public override IconUsage? Icon => FontAwesome.Solid.Calculator;
+        public override ModType Type => ModType.DifficultyIncrease;
+        public override double ScoreMultiplier => 1.0;
+        public override Type[] IncompatibleMods => base.IncompatibleMods.Append(typeof(ModEasyWithExtraLives)).Append(typeof(ModPerfect)).ToArray();
+
+        [SettingSource("Minimum accuracy", "Trigger a failure if your accuracy goes below this value.", SettingControlType = typeof(SettingsSlider<double, PercentSlider>))]
+        public BindableNumber<double> MinimumAccuracy { get; } = new BindableDouble
+        {
+            MinValue = 0,
+            MaxValue = 1,
+            Precision = 0.01,
+            Default = 0.9,
+            Value = 0.9,
+        };
+
+        private double baseScore;
+        private double maxBaseScore;
+        private double accuracy => maxBaseScore > 0 ? baseScore / maxBaseScore : 1;
+
+        protected override bool FailCondition(HealthProcessor healthProcessor, JudgementResult result)
+        {
+            if (!result.Type.IsScorable() || result.Type.IsBonus())
+                return false;
+
+            baseScore += result.Type.IsHit() ? result.Judgement.NumericResultFor(result) : 0;
+            maxBaseScore += result.Judgement.MaxNumericResult;
+
+            return accuracy < MinimumAccuracy.Value;
```

Comment:
Ideally this mod should not be locally computing accuracy, it should be using values already available elsewhere. Likely the easiest way to go is to implement `IApplicableToScoreProcessor` and bind to `ScoreProcessor.Accuracy` instead, which has the value ready. Not sure whether that will work, but it should be tried.

At the worst this should reuse pieces of `ScoreProcessor`.

Reply:
`ScoreProcessor.Accuracy` isn't updated until after `FailCondition` is called, so the accuracy known to the mod would always lag behind by one judgement.

Could you clarify what you mean by reusing pieces of `ScoreProcessor`? This logic is lifted [from here](https://github.com/ppy/osu/blob/0d7792da5003ecdac3e0ad05f90ae4fa1efca0b4/osu.Game/Rulesets/Scoring/ScoreProcessor.cs#L230-L238). We can't use `ScoreProcessor.TotalScore` to eliminate the need for the score attributes, as it may contain bonus score components. Do you want `ScoreProcessor` to expose a new method?

Reply:
Arbitrary failure triggering would definitely be useful. I'll look into how that would work.

Comment added.

Reply:
Here's the proposed fix. I think it's super ugly.

```diff
diff --git a/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs b/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs
index 732f54e356..fe81eacdcb 100644
--- a/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs
+++ b/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs
@@ -12,10 +12,11 @@
 using osu.Game.Rulesets.Scoring;
 using osu.Game.Rulesets.Judgements;
 using osu.Game.Scoring;
+using osu.Game.Screens.Play;
 
 namespace osu.Game.Rulesets.Mods
 {
-    public class ModAccuracyChallenge : ModFailCondition, IApplicableToScoreProcessor
+    public class ModAccuracyChallenge : ModFailCondition, IApplicableToScoreProcessor, IApplicableToPlayer
     {
         public override string Name => "Accuracy Challenge";
 
@@ -44,11 +45,10 @@ public class ModAccuracyChallenge : ModFailCondition, IApplicableToScoreProcesso
         };
 
         private ScoreProcessor scoreProcessor = null!;
+        private Player player = null!;
 
-        public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
-        {
-            this.scoreProcessor = scoreProcessor;
-        }
+        public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor) => this.scoreProcessor = scoreProcessor;
+        public void ApplyToPlayer(Player player) => this.player = player;
 
         public ScoreRank AdjustRank(ScoreRank rank, double accuracy) => rank;
 
@@ -58,7 +58,17 @@ protected override bool FailCondition(HealthProcessor healthProcessor, Judgement
             if (!result.Type.IsScorable() || result.Type.IsBonus())
                 return false;
 
-            return scoreProcessor.Accuracy.Value < MinimumAccuracy.Value;
+            // Must calculate inline as ScoreProcessor.Accuracy may not be up-to-date at this point in time.
+            var score = player.Score.ScoreInfo.DeepClone();
+
+            if (score.Statistics.ContainsKey(result.Type))
+                score.Statistics[result.Type]++;
+            else
+                score.Statistics[result.Type] = 1;
+
+            double currentAccuracy = scoreProcessor.ComputeAccuracy(score);
+
+            return currentAccuracy < MinimumAccuracy.Value;
         }
     }
 
```

I've offered an alternative at #22249 and committed changes to work with that.

Reply:
Swapping the processing order seems like the least bad option to me. I'm perfectly fine with whatever change helps move this forward. Thanks for the patch.